build: Make Dockerfile cacheable

underyx opened the pull request:
PR checklist:

- [ ] Documentation is up-to-date
- [ ] Changelog is up-to-date
- [ ] Change has no security implications (otherwise, ping security team)

underyx committed:
Add .dockerfile generating pre-commit hook

underyx committed:
build: Make Dockerfile cacheable

underyx committed:
Fix dev image push target

underyx committed:
Fix semgrep.bin not being present in docker

underyx committed:
fix(scan): Error on /src missing in docker only on no target

underyx committed:
Add more python versions to test matrix

underyx committed:
Cache core builds

underyx committed:
Clean up tests workflow file

underyx committed:
Remove make-version step

underyx committed:
Fix ocamlbuild cache key

underyx committed:
Merge build-tests into tests

underyx committed:
Fix core test job

underyx committed:
Fix python versions in tox.ini

underyx committed:
Enable dune cache in github actions

underyx committed:
Fix changelog reminder workflow

underyx committed:
Fix home stuff

underyx committed:
Keep messin with dune cache

underyx committed:
Add fetchjobs setting for submodule checkout

underyx committed:
Actually, try fetch.parallel

underyx committed:
Bump pipenv and tox

github-actions[bot] committed:
Update pytest snapshots

brendongo commented on .github/workflows/build-tests.yml:
```diff
@@ -1,107 +0,0 @@
-name: build-tests
```
what happened to these?

underyx replied:
moved to tests.yml!

- the events were essentially the same anyway
- I got to rename these jobs to match the same naming scheme as tests.yml
- the workflow diagram is also easier to understand now that all jobs are on the same page
- this will let us consolidate the dependencies a bit nicer going forward

underyx committed:
Finding a home here is more difficult than in SF

brendongo commented on semgrep/semgrep/commands/scan.py:
```diff
@@ -766,11 +764,9 @@ def scan(
 
     # change cwd if using docker
-    try:
+    if not targets:
```
is adjust_for_docker guaranteed to succeed if targets is not null?

underyx replied:
no; but the try-except here was just re-raising the exception anyway! i.e. we always wanted to crash if that function fails

brendongo commented on semgrep/semgrep/commands/scan.py:
```diff
@@ -819,19 +815,17 @@ def scan(
     # 'managed_output'. Output depends on file contents so we cannot have
     # already deleted the temporary stdin file.
-    with converted_pipe_targets(target_sequence) as target_sequence:
+    with converted_pipe_targets(targets) as targets:
```
seems iffy to shadow the variable like this

underyx replied:
yeah I was scared to change the logic though in case we were previously somehow depending on the shadowing :/

brendongo replied:
fair. can fix later if ever

underyx committed:
Limit cached .opam directories a bit

underyx committed:
refactor(target_manager): Remove usage of contextlib

The main reason is that this made our tests fail due to a py3.11 bug.
But I think this makes the code nicer anyway?

underyx committed:
Use docker build cache

underyx committed:
Fix checkouts in docker build jobs

brendongo commented on Dockerfile:
```diff
@@ -99,5 +86,12 @@ RUN chmod +x /entrypoint.sh
 
+COPY --from=semgrep-core /semgrep/semgrep-core/_build/default/src/cli/Main.exe /usr/local/bin/semgrep-core
+
+ENV SEMGREP_IN_DOCKER=1 \
```
why does this need to be set at the end instead of with all the other env vars?

underyx replied:
I do it so that we can decide to mess with these env vars and reuse all of the build cache from before

aryx commented:
@underyx can you explain what it means to mahe the "Dockerfile cacheable" and what are the implications
for people who want to modify the Dockerfile. Is there things they should avoid?
